[FSDP] Fix input grad propagation when using param mixed precision

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90921 [FSDP] Fix input grad propagation when using param mixed precision**
* #90858 [FSDP][Easy] ufmt files
* #90840 [FSDP][BE] Remove `_module_to_handles`, `HandleConfig`; use term "fqn"; clarify docs

For parameter mixed precision, we cast the inputs to the low precision parameter dtype. If the input has tensors that require gradient, then we must cast them in place in order for them to receive a gradient. The cast should be tracked by autograd (e.g. with `grad_fn` equal to `ToCopyBackward0`). This removes the `torch.no_grad` context when calling `_apply_to_tensors`.

Commit:
[FSDP] Fix input grad propagation when using param mixed precision

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix input grad propagation when using param mixed precision"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix input grad propagation when using param mixed precision"




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -456,12 +456,13 @@ def _cast_fp_inputs_to_dtype(
             return x
-        y = x.to(dtype)
-        # Explicitly copy over `requires_grad` since this runs inside
-        # `torch.no_grad()`
```
I think this comment about `_cast_fp_inputs_to_dtype` running in `no_grad` is outdated given my check just now.

Commit:
Update on "[FSDP] Fix input grad propagation when using param mixed precision"


For parameter mixed precision, we cast the inputs to the low precision parameter dtype. If the input has tensors that require gradient, then we must cast them in place in order for them to receive a gradient. Otherwise, the tensor that resulted from the out-of-place cast receives the gradient and is not in scope to the user. To preserve BC as much as possible, this PR only does the in-place cast if the tensor requires gradient.

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -456,15 +456,9 @@ def _cast_fp_inputs_to_dtype(
-
-    with torch.no_grad():
-        return (_apply_to_tensors(cast_fn, args), _apply_to_tensors(cast_fn, kwargs))
+        return x.to(dtype)
```
hmm, I am a bit concerned about unforseen BC issues, but also our testing surface is quite solid. Did we consider doing this only for inputs that require grad to err on the safe side?

awgu replied:
I think if `x` does not require gradient, then `x.to(dtype)` will not be tracked by autograd. if `x` does require gradient, then `x.to(dtype)` will be tracked by autograd. This should be exactly the behavior we want. In other words, the casing should be handled naturally already if I understand correctly.

rohan-varma commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -718,6 +718,50 @@ def never_wrap_policy(*args, **kwargs):
+        x_float = torch.randn(32, 1024, device="cuda", dtype=torch.float32)
+        x_float.requires_grad_()
+        fsdp_model(x_float).sum().backward()
+        self.assertTrue(x_float.grad is not None)
```
nit: should we also assert on the dtype?

awgu replied:
Good point.

rohan-varma commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -718,6 +718,50 @@ def never_wrap_policy(*args, **kwargs):
+        )
+        # Use an input with dtype not equal to the mixed precision
+        # `param_dtype` so that it gets cast
+        x_float = torch.randn(32, 1024, device="cuda", dtype=torch.float32)
```
nit: you can pass `requires_grad=True` and avoid the set below.

rohan-varma reviewed (approved):
LGTM, thanks for the fix!

Commit:
Update on "[FSDP] Fix input grad propagation when using param mixed precision"


For parameter mixed precision, we cast the inputs to the low precision parameter dtype. If the input has tensors that require gradient, then we must cast them in place in order for them to receive a gradient. The cast should be tracked by autograd (e.g. with `grad_fn` equal to `ToCopyBackward0`). This removes the `torch.no_grad` context when calling `_apply_to_tensors`.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
